src/mlpack/bindings/cli/add_to_po.hpp

```diff
@@ -2,83 +2,86 @@
  * @file add_to_po.hpp
  * @author Ryan Curtin
  *
- * Utility functions to add options to boost::program_options based on their
+ * Utility functions to add options to CLI11 based on their
  * type.
  *
  * mlpack is free software; you may redistribute it and/or modify it under the
  * terms of the 3-clause BSD license.  You should have received a copy of the
  * 3-clause BSD license along with mlpack.  If not, see
  * http://www.opensource.org/licenses/BSD-3-Clause for more information.
  */
-#ifndef MLPACK_BINDINGS_CLI_ADD_TO_PO_HPP
-#define MLPACK_BINDINGS_CLI_ADD_TO_PO_HPP
+#ifndef MLPACK_BINDINGS_CMD_ADD_TO_PO_HPP
+#define MLPACK_BINDINGS_CMD_ADD_TO_PO_HPP
 
 #include <mlpack/core/util/param_data.hpp>
-#include <boost/program_options.hpp>
 #include <mlpack/core/util/is_std_vector.hpp>
 #include "map_parameter_name.hpp"
 
+#include <CLI/CLI.hpp>
+
 namespace mlpack {
 namespace bindings {
-namespace cli {
+namespace cmd {
 
 /**
- * Add a non-vector option to boost::program_options.
+ * Add a non-vector option to CLI11.
  *
- * @param boostName The name of the option to add to boost::program_options.
+ * @param boostName The name of the option to add to CLI11.
  * @param descr Description string for parameter.
  * @param desc Options description to add parameter to.
  */
 template<typename T>
-void AddToPO(const std::string& boostName,
+void AddToPO(const std::string& cliName,
              const std::string& descr,
-             boost::program_options::options_description& desc,
+             CLI::App& app,
              const typename boost::disable_if<util::IsStdVector<T>>::type* = 0,
              const typename boost::disable_if<std::is_same<T, bool>>::type* = 0)
 {
-  desc.add_options()(boostName.c_str(), boost::program_options::value<T>(),
-      descr.c_str());
+  T value;
+  app.add_option(cliName.c_str(), value, descr.c_str());
```

Comment:
I was debugging the segfault that happens during parsing, which seems to occur in `parse_command_line.hpp`.  During this I started to read about CLI11 and see what its way of setting variables is.  It seems like typically you would call `add_option()` just like this, and when parsing it will set the value of the given variable.  But in this case, we've passed a temporary instead of the actual value, so I think this probably causes some problems.

Ideally, I guess we could have CLI11 parse *directly* into `ParamData::value`, but that seems hard because it's a `boost::any`.  However, I see that we don't just have to pass a value:

https://github.com/CLIUtils/CLI11#option-types

We can also pass a lambda function!  So I wonder if something like this will work:

```
app.add_option_function(cliName.c_str(), [paramData](T& t) { paramData.value = value; paramData.wasPassed = true; });
```

or... something like that.  I hope what I have written makes sense.

I guess, to clarify and rewrite: currently, the structure we were doing with boost::program_options included this `AddToPO()` function, which I believe was called in the first part of `ParseCommandLine()`.  Then we would use `boost::program_options` to do the parsing, and then iterate over the returned values and call `SetParam()` to set their values in each `ParamData` struct.  What I am suggesting here is that we don't need to do that two-step process---instead, in this `AddToPO()` function (which probably we could rename), we could give CLI11 a function pointer that would do the same functionality as `SetParam()` for each type (and set `wasPassed`), and then there is no need for a second loop over all of the parsed options, since CLI11 has set all the values correctly during parsing.

Reply:
Please, let me know if I can clarify the idea.  I don't know how coherently I've written it down. :+1:

Reply:
It does make sense to me. I will try to implement it. I was only trying to adapt the code. I was able to understand it.
Actually what I found hard in mlpack is to understand why the code is written this way, rather what does the code do.

Reply:
Hehe, yeah, the way it is written is fairly convoluted.  It is weird that everything has to be accessed through this `functionMap`, like `CLI::GetSingleton().functionMap["SetParam"](...)` and then the parameters have to be casted to void pointers or similar.

The reason for this is that `ParamData` struct does not have any ability to store any type information (it's not templatized, and if it was we wouldn't be able to hold `std::vector<ParamData<T>>` or similar since `T` can be different for every type).  Therefore we *only* have access to the type information exactly at the point of instantiation of the `CLIOption` (or `PyOption` or `GoOption`, etc.).  Thus we have to make function pointers to the correctly-typed functions like `SetParam`, `AddToPO`, etc. in order to be able to have different behavior depending on the type (which you can see we do need in the various functions like `SetParam()`, etc.).

Hopefully that makes sense; let me know if I can clarify. :+1:

Reply:
@rcurtin You agree if I use lambda function with `paramData` we will have to remove the const from the function pointer:
https://github.com/mlpack/mlpack/blob/a147cfbddcb7b8c238c10a763659080d782883e5/src/mlpack/core/util/cli.hpp#L300-L302
Is there any downside to that? Because we will have to reiterate throughout the entire code base for all functions called by `GetSingletin()`

Reply:
That seems fine to me.  I suppose there's no way around it, unless you want to do a `const_cast<>` or something to turn the `const util::ParamData&` into a `util::ParamData&` inside of `AddToPO()`.

Reply:
@rcurtin  Basically I had to remove `const` from most of the places that use `util::ParamData` and still there somewhere a hidden `const` that I can not see.
I can not find anywhere the usage of `boost_any` with `ParamData` I am sure I did not read the code perfectly. 
```
/meta/mlpack/src/mlpack/bindings/cli/add_to_po.hpp: In instantiation of ‘void mlpack::bindings::cmd::AddToPO(const string&, mlpack::util::ParamData&, CLI::App&, typename boost::disable_if<ml
pack::util::IsStdVector<T> >::type*, typename boost::disable_if<std::is_same<T, bool> >::type*) [with T = std::__cxx11::basic_string<char>; std::string = std::__cxx11::basic_string<char>; ty
pename boost::disable_if<mlpack::util::IsStdVector<T> >::type = void; typename boost::disable_if<std::is_same<T, bool> >::type = void]’:
/meta/mlpack/src/mlpack/bindings/cli/add_to_po.hpp:111:79:   required from ‘void mlpack::bindings::cmd::AddToPO(mlpack::util::ParamData&, const void*, void*) [with T = std::__cxx11::basic_st
ring<char>]’
/meta/mlpack/src/mlpack/bindings/cli/cli_option.hpp:152:55:   required from ‘mlpack::bindings::cmd::CMDOption<N>::CMDOption(N, const string&, const string&, const string&, const string&, boo
l, bool, bool, const string&) [with N = std::__cxx11::basic_string<char>; std::string = std::__cxx11::basic_string<char>]’
/meta/mlpack/src/mlpack/bindings/cli/parse_command_line.hpp:28:1:   required from here
/meta/mlpack/src/mlpack/bindings/cli/add_to_po.hpp:41:17: error: passing ‘const boost::any’ as ‘this’ argument discards qualifiers [-fpermissive]
   41 |     param.value = value;
      |     ~~~~~~~~~~~~^~~~~~~
In file included from /meta/mlpack/src/mlpack/core/util/cli.hpp:22,
                 from /meta/mlpack/src/mlpack/methods/pca/pca_main.cpp:14:
/usr/include/boost/any.hpp:122:15: note:   in call to ‘boost::any& boost::any::operator=(ValueType&&) [with ValueType = const std::__cxx11::basic_string<char>&]’
  122 |         any & operator=(ValueType&& rhs)
      |               ^~~~~~~~
In file included from /meta/mlpack/src/mlpack/bindings/cli/cli_option.hpp:20,
                 from /meta/mlpack/src/mlpack/core/util/mlpack_main.hpp:37,
                 from /meta/mlpack/src/mlpack/methods/pca/pca_main.cpp:15:
/meta/mlpack/src/mlpack/bindings/cli/add_to_po.hpp:42:21: error: assignment of member ‘mlpack::util::ParamData::wasPassed’ in read-only object
   42 |     param.wasPassed = true;
      |     ~~~~~~~~~~~~~~~~^~~~~~
/meta/mlpack/src/mlpack/bindings/cli/add_to_po.hpp:40:17: error: no matching function for call to ‘CLI::App::add_option(const char*, mlpack::bindings::cmd::AddToPO<std::__cxx11::basic_string
<char> >::<lambda(const std::__cxx11::basic_string<char>&)>, const char*)’
   40 |   app.add_option(cliName.c_str(), [param](const T& value){
      |   ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   41 |     param.value = value;
      |     ~~~~~~~~~~~~~~~~~~~~
   42 |     param.wasPassed = true;
      |     ~~~~~~~~~~~~~~~~~~~~~~~
   43 |   }, param.desc.c_str());
      |   ~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/CLI/CLI.hpp:31,
```

Reply:
I resolved this error, I forget to pass by reference in lambda